### ISSUE TITLE
Two different references to eslint

### ISSUE THREAD

**Issue by btmills** [eslint/eslint-tester#24]
Split from #22.

First, @PiPeep:

To add to what @zpao said, in #21, there appears to be a line saying:

```js
...
validator = require("eslint").validator,
...
```

However, [`ESLintTester` is already be passed `eslint.linter`](https://github.com/eslint/eslint-tester/blob/18d2a86dd45c67ebc7fb184a365229b6377e8ffb/tests/lib/eslint-tester.js#L54). IMHO, either it should be able to pick validator out from `eslint.linter`, or require we pass `eslint` instead of `eslint.linter`.

Right now, it's possible that eslint-tester might end up with two different versions of eslint, and cause some strange issues.

Then @btmills:

> either it should be able to pick validator out from eslint.linter, or require we pass eslint instead of eslint.linter

Agreed. Unfortunately, due to the complicated relationship between eslint and eslint-tester, either option will mean handling two concurrent releases again. @nzakas or @ilyavolodin - thoughts?

**Option 1:** Export `validator` from `lib/eslint`, and it no longer needs to be exported from `lib/api`.

**Option 2:** Rules should pass `lib/api`, not `lib/eslint` to eslint-tester.

Architecturally, I think it's more proper that eslint-tester would use the public API (so option 2), but that would require changing *every single rule*, and we can't suddenly break third-party plugin rules, meaning we might have to go with option 1.

**Comment by nzakas** [comment]
Looking this over again, we have a weird relationship between ESLint and the tester. ESLint has the tester as a dev dependency. If the tester has ESLint as a peer dependency, that would result in a second installation (?). So the peer dependency makes sense for all projects outside of ESLint, but since ESLint is the primary consumer, that doesn't make sense.

We probably want to move to a model where the ESLint package must be passed into the tester rather than being included on its own. That would eliminate the weird dependency issue and be more architecturally sound.

We would have to change all our tests, but I've done that fairly easily using a perl script. 

We can also do a major version bump for such a change, so all existing uses will remain working until they manually upgrade.

**Comment by nzakas** [comment]
An interesting thought occurred to me: maybe we are just fighting the tide here and we should merge tester back into ESLint. Doing so would solve nearly all of the problems we've been discussing, and would make ESLint itself a completely standalone solution for dev and testing.

**Comment by btmills** [comment]
I wasn't around when the split originally occurred, so I don't know why that happened in the first place. What would be the downsides to putting it back together? As far as upsides, being able to avoid managing parallel releases and the `npm link` sequence we use now would be pretty big wins.

**Comment by nzakas** [comment]
The downsides are (IMO):

* Unclear dependency on Mocha now (should ESLint have it as a dependency?)
* Can't install just the tester, always need to install ESLint (maybe problematic for custom rules?)
* Combining issue trackers can lead to tester issues losing visibility/focus

Maybe these issues aren't bad enough to not merge it back in. 

I can't find a record of the rational for splitting, as the project was still quite young (7 months old) when it happened.

**Comment by btmills** [comment]
@nzakas do you have any leanings here as to the correct solution? I can't convince myself either way, but neither do I think I've wrapped my head around all the issues at play.

**Comment by ilyavolodin** [comment]
I think I would also vote for merging tester back into eslint. I really don't see what we are gaining by having it separate. If I'm not mistaken it was extracted right before we introduced plugins, and I can imagine that we though that having it as a separate package would help test plugins more easily. Problem is, all of the plugins still need to include eslint as dependency anyways right now, so having it as a stand-alone solution doesn't help.

**Comment by nzakas** [comment]
Yeah, I'd say we merge back in. But, I think while we do that, we should take a hard look at the API and see what changes would make sense. Some of the stuff makes it hard for others to use:

1. Tied to Mocha
2. Implicit loading of rules

Would be nice to clean that up.

**Comment by lo1tuma** [comment]
If we merge eslint-tester back into eslint we have to make mocha and chai dependencies instead of devDependencies. I’m not sure if we want to add these additional dependencies for all users of eslint.

**Comment by bgw** [comment]
@lo1tuma see my stance on this in issue #25.

**Comment by nzakas** [comment]
@lo1tuma these are some of the issues with the first point I mentioned. We could, in theory, create a rule tester that doesn't need to know about Mocha or Chai ( we could use native assert module ). We definitely don't need `describe` baked in, we could just have people do that themselves. 

This is just off the top of my head, but what if we did something like this:

```js
var RuleTester = require("../../util/rule-tester"),
    linter = require('../../eslint'),
    rule = require('../../../rules/my-rule');

describe('my-rule', function() {

    var tester = new RuleTester(linter, rule);

    // optionally, say you want to use "it"
    tester.separateWith(it);

    tester.add({ /* valid/invalid */);

});
```
